### PR TITLE
[cling] Diagnose shadowing of decls in the `std` namespace

### PR DESCRIPTION
**Pull request by jalopezg-r00t**
Related to the discussion in [ROOT-5971](https://sft.its.cern.ch/jira/browse/ROOT-5971), this commit improves the situation by diagnosing the shadowing of declarations in the `std` namespace.

In summary, given that ROOT issues an implicit `using namespace std;` and due to the way `SemaLookup` works, decls that shadow something in the `std` namespace cannot be referenced without using their fully-qualified name, i.e. `::xxx`.
See discussion in ROOT-5971 for the details.  This pull-request improves the situation by emitting a warning for that case.

## Changes or fixes:
- Add `Interpreter::m_InputFlags`, which provides additional information about the context in which parsing occurs.  In particular, the `kInputFromFile` and `kIFFLineByLine` flags allow to tell whether a sequence of calls to `Interpreter::process()` is issued from `MetaProcessor::readInputFromFile()` for the contents of an external file.
- Update cling's DefinitionShadower to emit a warning if a new declaration shadows a name in the `std` namespace (that became available via the implicit `using namespace std`).  This diagnostic is disabled for unnamed macros, as input is typically ingested in a single `process()` call, which makes all the variables local to the generated wrapper. 

This requires to change the lookup redeclaration kind to `NotForRedeclaration`, which has a performance penalty, given that the lookup will follow using directives, etc.  This is approximately one order of magnitude slower than `ForVisibleRedeclaration` (see table below).  The overhead seems acceptable given that this kind of lookup already occurs when a declaration is referenced and that definition is rare w.r.t. use.
```
+-------------------------+-------------------------------+
| Redeclaration kind      | Time (us) for 1000000 lookups |
+-------------------------+-------------------------------+
| ForVisibleRedeclaration |                         27617 |
| NotForRedeclaration     |                        717458 |
+-------------------------+-------------------------------+
```

## Checklist:
- [X] tested changes locally

This PR improves the situation for [ROOT-5971](https://sft.its.cern.ch/jira/browse/ROOT-5971).

### CODE REVIEW

File: interpreter/cling/lib/Interpreter/DefinitionShadower.cpp [hunk 1]
```diff
@@ -132,6 +140,18 @@ namespace cling {
...
       D->setInvalidDecl();
+
+    // Diagnose shadowing of decls in the `std` namespace (see ROOT-5971).
+    // For unnamed macros, the input is ingested in a single `Interpreter::process()`
+    // call. Do not emit the warning in that case, as all references are local
+    // to the wrapper function and this diagnostic might be misleading.
+    if (shadowsDeclInStd
+        && ((m_Interp.getInputFlags() & (Interpreter::kInputFromFile
+                                         | Interpreter::kIFFLineByLine))
+             != Interpreter::kInputFromFile)) {
```
Comment (hahnjo) [inline interpreter/cling/lib/Interpreter/DefinitionShadower.cpp:151]: This means we also warn for `m_Interp.getInputFlags() == 0`, right? Does that mean all of REPL and things like cppyy and RDF?

Comment (jalopezg-r00t) [reply interpreter/cling/lib/Interpreter/DefinitionShadower.cpp:151]: Yes, that is right.  It means that we warn if definition shadowing did its job (that typically implies that we entered via `Interpreter::process()`), and that the user was trying to shadow a global declaration that conflicts with something we pulled in to the global scope from `std` (via the implicit `using namespace std;`).

In other words, this happens for REPL, but not for cppyy/RDF as they generate wrapper functions that should not be touched by DefinitionShadower. See below for an example.

```c++
root [0] unsigned data = 0;
ROOT_prompt_0:1:1: warning: 'data' shadows a declaration with the same name in the 'std' namespace; consider using '::data' to reference this declaration
unsigned data = 0;
^
root [1] data
ROOT_prompt_1:1:1: error: reference to 'data' is ambiguous
data
^
ROOT_prompt_0:1:10: note: candidate found by name lookup is '__cling_N50::data'
unsigned data = 0;
         ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:319:5: note: candidate found by name lookup is 'std::data'
    data(initializer_list<_Tp> __il) noexcept
    ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:290:5: note: candidate found by name lookup is 'std::data'
    data(_Container& __cont) noexcept(noexcept(__cont.data()))
    ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:300:5: note: candidate found by name lookup is 'std::data'
    data(const _Container& __cont) noexcept(noexcept(__cont.data()))
    ^
/usr/lib/gcc/x86_64-pc-linux-gnu/11.1.0/../../../../include/c++/11.1.0/bits/range_access.h:310:5: note: candidate found by name lookup is 'std::data'
    data(_Tp (&__array)[_Nm]) noexcept
    ^
root [2] ::data
(unsigned int) 0
```

**Review by hahnjo — APPROVED** [review]
Thanks for clarifying. LGTM, I think the last test results were unrelated.

File: interpreter/cling/lib/Interpreter/DefinitionShadower.cpp [hunk 2]
```diff
@@ -94,9 +98,12 @@ namespace cling {
   }
 
   void DefinitionShadower::invalidatePreviousDefinitions(NamedDecl *D) const {
+    // NotForRedeclaration: lookup anything visible; follows using directives
     LookupResult Previous(*m_Sema, D->getDeclName(), D->getLocation(),
-                   Sema::LookupOrdinaryName, Sema::ForVisibleRedeclaration);
+                   Sema::LookupOrdinaryName, Sema::NotForRedeclaration);
+    Previous.suppressDiagnostics();
     m_Sema->LookupQualifiedName(Previous, m_TU);
+    bool shadowsDeclInStd{};
```
Comment (Axel-Naumann) [inline interpreter/cling/lib/Interpreter/DefinitionShadower.cpp:106]: Could you use the more explicit version, please?
```suggestion
    bool shadowsDeclInStd = false;
```

**Review by Axel-Naumann — APPROVED** [review]
Some minor possible improvements (see comments), but LGTM!

File: interpreter/cling/lib/Interpreter/DefinitionShadower.cpp [hunk 3]
```diff
@@ -116,6 +123,7 @@ namespace cling {
                                 /*IsForUsingDecl=*/false))
         continue;
 
+      shadowsDeclInStd |= Prev->isInStdNamespace();
```
[the remark this answers is not in the document]
Comment (jalopezg-r00t) [reply interpreter/cling/lib/Interpreter/DefinitionShadower.cpp:126]: We actually don't require that :-).  We are performing a lookup in the global scope; if we see something that is in the `std` namespace, there must have been a prior using directive that made that visible in the TU.

Just to make sure:
```c++
[cling]$ gClingOpts->AllowRedefinition = 1
(bool) true
[cling]$ unsigned data = 1;
[cling]$ // no diagnostic emitted
[cling]$ .q
```

**Pull request merged**